docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+In addition to that we have to change the `PodAntiAffinity` rule.
+The following would be an example of the `affinity` term for a Pod.
+The example assumes that process number modulo `DesiredFaultDomains` is `0` and we have set `BinPackAll`:
+
+```yaml
+  affinity:
+    # Affinity to schedule Pods with the same distribution key together
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+              key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across fault domains.
+If `Required` is set to false this will be on a best effort base and can be useful if a fault domain is full but the Pod should be scheduled anyway.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
+    // DesiredFaultDomains desfines the number of desired fault domain.
+    // Should be greater than 0 is fault domain distribution is enabled.
+    // Default: 0
+    DesiredFaultDomains *int
+    // BinPack defines what processes should be bin packed to the fault domains.
+    // Default: "Class"
+    BinPack *BinPack
+    // Required defines if the affinity terms are required or preffered.
+    // Default: false
+    Required *bool
+}
+```
+
+The operator won't guarantee that the fault domains are equally used.
+Depending on the replacements for unreachable Pods we could use one fault domain more than another (or depending on the capacity),
+We also don't guarantee that we use exactly the number of `DesiredFaultDomains` e.g. when we spawn less processes than `DesiredFaultDomains` or when a fault domain is out of capacity and we are not using the `Required` with `true`.
```

Comment:
If the `Required` setting is true and a fault domain is out of capacity, I would expect that we would automatically spread to another fault domain. Is that correct?

Reply:
That's not correct. In that case the Pod would stay in pending and is not able to be scheduled. Otherwise the required setting would be a little bit meaningless. The required setting only defines if we use a `preferredDuringSchedulingIgnoredDuringExecution` or a `requiredDuringSchedulingIgnoredDuringExecution` with the latter and a fault domain out of resources the Pod is stuck in Pending until resources will be available.

Reply:
Does the `required` setting affect the constraint type for the affinity toward the same logical fault domain, the anti-affinity against a different logical fault domain, or both?

Reply:
My assumption was that the `required` setting would only affect the anti-affinity rules, and that the affinity rules would always be preferred in order to prevent running out of capacity.

Reply:
Updated.